### PR TITLE
ENH: Generate random rotations

### PR DESCRIPTION
**Pull request by adbugger**
Used `scipy.stats.special_ortho_group` to generate random rotations.

### CODE REVIEW

**Comment by adbugger** [comment]
There seems to be an import error with `from scipy.stats import special_ortho_group`. What is the alternative?

**Comment by nmayorov** [comment]
My suggestion is to generate 4 normally distributed numbers as quaternion components and then simply normalize them. It gives uniform random rotations.

File: scipy/spatial/transform/rotation.py [hunk 1]
```diff
@@ -872,3 +874,31 @@ def __getitem__(self, indexer):
                   array.
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def random(cls, num=None, state=None):
+        """Generate rotations uniformly distributed in 3D space.
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I don't think mentioning 3D space is correct here. I think "distributed uniformly" would be sufficient (intuitively it's clear what it means).

File: scipy/spatial/transform/rotation.py [hunk 2]
```diff
@@ -872,3 +874,31 @@ def __getitem__(self, indexer):
...
+
+        Parameters
+        ----------
+        num : None or int
+            Number of random rotations to generate. If `None`, then a single
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: If None (default), ...

Also I suggest not to use single backticks for things like None, they are to reference arguments.

File: scipy/spatial/transform/rotation.py [hunk 3]
```diff
@@ -872,3 +874,31 @@ def __getitem__(self, indexer):
...
+            rotation is generated. Default is `None`.
+        state : None or `numpy.random.RandomState` object
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: random_state

File: scipy/spatial/transform/rotation.py [hunk 4]
```diff
@@ -872,3 +874,31 @@ def __getitem__(self, indexer):
...
+            Object for manually setting the generator state.
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I don't particularly like this description, not sure how to improve it.

File: scipy/spatial/transform/rotation.py [hunk 5]
```diff
@@ -872,3 +874,31 @@ def __getitem__(self, indexer):
...
+
+        Returns
+        -------
+        rotation : `Rotation` instance
+            Contains a single rotation if `num` is `None`. Otherwise contains a
+            stack of `num` rotations.
+        """
+        if state is not None:
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: You should handle it in a different way, definitely not setting the global state.

Check how random state is handled in this function https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/utils/validation.py#L617

Eventually you need to call `random_state.normal`.

File: scipy/spatial/transform/rotation.py [hunk 6]
```diff
@@ -872,3 +874,31 @@ def __getitem__(self, indexer):
...
+            np.random.set_state(state)
+        size = 1 if num is None else num
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I think writing it as 
```
if num is None:
   sample = np.random.normal(size=4)
else:
   sample = np.random.normal(size=(num, 4))
```
is cleaner.

File: scipy/spatial/transform/rotation.py [hunk 7]
```diff
@@ -210,6 +211,7 @@ class Rotation(object):
     __mul__
     inv
     __getitem__
+    random
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Should we put it somewhere in factory methods?

**Review by nmayorov — COMMENTED** [review]
After considering my suggestions, I think it is good to be merged.

File: scipy/spatial/transform/rotation.py (hunk 7 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: -0.5. I don't think it is appropriate. Here the main goal is random sampling, not initialization. What do you suggest?
